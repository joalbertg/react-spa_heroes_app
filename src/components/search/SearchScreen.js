import React, { useMemo } from 'react';
import { useLocation } from 'react-router-dom';
import queryString from 'query-string';

import { HeroCard } from '../heroes/HeroCard';
import { useForm } from '../../hooks/useForm';
import { getHeroByName } from '../../selectors/getHeroByName';

export const SearchScreen = ({ history }) => {
  const location = useLocation();
  const { q = '' } = queryString.parse(location.search);
  const [formValues, handleInputChange] = useForm({
    searchText: q
  });

  const { searchText } = formValues;
  const heroesFiltered = useMemo(() => getHeroByName(q), [q]);
  const handleSearch = event => {
    event.preventDefault();
    if(searchText.trim().length < 3) return;

    history.push(`?q=${searchText}`);
  }

  return(
    <div className='row'>
      <div className='col-5'>
        <h4>Search</h4>
        <hr />

        <form onSubmit={handleSearch}>
          <input
            type='text'
            name='searchText'
            placeholder='Find your favorite hero'
            className='form-control'
            autoComplete="off"
            value={searchText}
            onChange={handleInputChange}
          />
          <button
            className='btn btn-outline-primary btn-block mt-1'
            type='submit'
          >
            Search...
          </button>
        </form>
      </div>
      <div className='col-7'>
        <h4>Results</h4>
        <hr />
        {
          (q === '')
            &&
            <div className="alert alert-info">Search a hero</div>
        }
        {
          (q !== '' && heroesFiltered.length === 0)
            &&
            <div className="alert alert-danger">There is no a hero with {q}</div>
        }
        {
          heroesFiltered.map(hero => (
            <HeroCard
              key={hero.id}
              hero={hero}
              className="animate__animated animate__fadeIn"
            />
          ))
        }
      </div>
    </div>
  );
}

