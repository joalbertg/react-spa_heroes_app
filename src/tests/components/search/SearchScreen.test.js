import React from 'react';
import { mount } from 'enzyme';
import {
  MemoryRouter,
  Route,
  useLocation
} from 'react-router-dom';

import { SearchScreen } from '../../../components/search/SearchScreen';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
     useLocation: jest.fn()
}));

describe('Tests SearchScreen component', () => {
  test('Should display correctly', () => {
    useLocation.mockReturnValueOnce({ search: '' });

    //const wrapper = mount(
    //  <MemoryRouter initialEntries={['/search']}>
    //    <Route path='/search' component={SearchScreen} />
    //  </MemoryRouter>
    //);
    const wrapper= mount(<SearchScreen />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.row > .col-5 > h4').text()).toBe('Search');
    expect(wrapper.find('input').exists()).toBeTruthy();
    expect(wrapper.find('.row .col-7 > h4').text()).toBe('Results');
    expect(wrapper.find('.alert.alert-info').text()).toBe('Search a hero');
  });

  test('Should return a batman card and set input with batman', () => {
    //const wrapper = mount(
    //  <MemoryRouter initialEntries={['/search?q=batman']}>
    //    <Route path='/search' component={SearchScreen} />
    //  </MemoryRouter>
    //);
    const hero = 'batman';
    const query = `q=${hero}`;
    useLocation.mockReturnValueOnce({ search: query });

    const wrapper= mount(
      <MemoryRouter>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input').prop('value')).toBe(hero);
    expect(wrapper.find('h5.card-title').text()).toBe('Batman');
  });

  test('Should call submit and return a superman card and set input with superman', () => {
    const hero = 'superman';
    const query = `q=${hero}`;
    const history = {
      push: jest.fn()
    }

    useLocation.mockReturnValueOnce({ search: query });

    const wrapper= mount(
      <MemoryRouter>
        <SearchScreen history={history} />
      </MemoryRouter>
    );

    wrapper.find('form').prop('onSubmit')({ preventDefault(){} });

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input').prop('value')).toBe(hero);
    expect(wrapper.find('h5.card-title').text()).toBe('Superman');
    expect(history.push).toHaveBeenCalled();
    expect(history.push).toHaveBeenCalledTimes(1);
    expect(history.push).toHaveBeenCalledWith('?' + query);
  });

  test('Should call submit and show error if hero not found', () => {
    const hero = 'anything';
    const query = `q=${hero}`;
    const history = {
      push: jest.fn()
    }

    useLocation.mockReturnValueOnce({ search: query });

    const wrapper= mount(
      <MemoryRouter>
        <SearchScreen history={history} />
      </MemoryRouter>
    );

    wrapper.find('form').prop('onSubmit')({ preventDefault(){} });

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input').prop('value')).toBe(hero);
    expect(wrapper.find('h5.card-title').exists()).toBeFalsy();
    expect(wrapper.find('.col-7 .alert.alert-danger').text()).toBe(`There is no a hero with ${hero}`);
    expect(history.push).toHaveBeenCalled();
    expect(history.push).toHaveBeenCalledTimes(1);
    expect(history.push).toHaveBeenCalledWith('?' + query);
  });
});

