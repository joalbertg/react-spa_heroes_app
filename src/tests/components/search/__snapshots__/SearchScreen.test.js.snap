// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests SearchScreen component Should call submit and return a superman card and set input with superman 1`] = `
<div
  className="row"
>
  <div
    className="col-5"
  >
    <h4>
      Search
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Find your favorite hero"
        type="text"
        value="superman"
      />
      <button
        className="btn btn-outline-primary btn-block mt-1"
        type="submit"
      >
        Search...
      </button>
    </form>
  </div>
  <div
    className="col-7"
  >
    <h4>
      Results
    </h4>
    <hr />
    <div
      className="card ms-3 animate__animated animate__fadeIn"
      style={
        Object {
          "maxWidth": 540,
        }
      }
    >
      <div
        className="row no-gutters"
      >
        <div
          className="col-md-4"
        >
          <img
            alt="Superman"
            className="card-img"
            src="./assets/heroes/dc-superman.jpg"
          />
        </div>
        <div
          className="card-md-8"
        >
          <div
            className="card-body"
          >
            <h5
              className="card-title"
            >
              Superman
            </h5>
            <p
              className="card-text"
            >
              Kal-El
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                Action Comics #1
              </small>
            </p>
            <a
              href="/hero/dc-superman"
              onClick={[Function]}
            >
              more...
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tests SearchScreen component Should call submit and show error if hero not found 1`] = `
<div
  className="row"
>
  <div
    className="col-5"
  >
    <h4>
      Search
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Find your favorite hero"
        type="text"
        value="anything"
      />
      <button
        className="btn btn-outline-primary btn-block mt-1"
        type="submit"
      >
        Search...
      </button>
    </form>
  </div>
  <div
    className="col-7"
  >
    <h4>
      Results
    </h4>
    <hr />
    <div
      className="alert alert-danger"
    >
      There is no a hero with 
      anything
    </div>
  </div>
</div>
`;

exports[`Tests SearchScreen component Should display correctly 1`] = `
<div
  className="row"
>
  <div
    className="col-5"
  >
    <h4>
      Search
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Find your favorite hero"
        type="text"
        value=""
      />
      <button
        className="btn btn-outline-primary btn-block mt-1"
        type="submit"
      >
        Search...
      </button>
    </form>
  </div>
  <div
    className="col-7"
  >
    <h4>
      Results
    </h4>
    <hr />
    <div
      className="alert alert-info"
    >
      Search a hero
    </div>
  </div>
</div>
`;

exports[`Tests SearchScreen component Should return a batman card and set input with batman 1`] = `
<div
  className="row"
>
  <div
    className="col-5"
  >
    <h4>
      Search
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Find your favorite hero"
        type="text"
        value="batman"
      />
      <button
        className="btn btn-outline-primary btn-block mt-1"
        type="submit"
      >
        Search...
      </button>
    </form>
  </div>
  <div
    className="col-7"
  >
    <h4>
      Results
    </h4>
    <hr />
    <div
      className="card ms-3 animate__animated animate__fadeIn"
      style={
        Object {
          "maxWidth": 540,
        }
      }
    >
      <div
        className="row no-gutters"
      >
        <div
          className="col-md-4"
        >
          <img
            alt="Batman"
            className="card-img"
            src="./assets/heroes/dc-batman.jpg"
          />
        </div>
        <div
          className="card-md-8"
        >
          <div
            className="card-body"
          >
            <h5
              className="card-title"
            >
              Batman
            </h5>
            <p
              className="card-text"
            >
              Bruce Wayne
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                Detective Comics #27
              </small>
            </p>
            <a
              href="/hero/dc-batman"
              onClick={[Function]}
            >
              more...
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
